---
title: "Introduction to the kmeRs package"
author:
- name: Rafal Urniaz, PhD^[BioTesseract Cambridge Bioinformatics Solutions, Cambridgeshire, Cambridge, UK]
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
  # pdf_document
  # md_document:
  #    variant: markdown_github
bibliography: citations.bib
vignette: >
  %\VignetteIndexEntry{Introduction to the kmeRs R package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
## Introduction

The package contains tools to calculate similarity score matrix for DNA k-mers. The pairwise similarity score is calculated using PAM or BLOSUM substitution matrix. The results are evaluated by similarity score calculated by Needleman-Wunsch (global) [@NEEDLEMAN1970443] or Smith-Waterman (local) alignment. [@SMITH1981195]. Higher similarity score indicates more similar sequences for BLOSUM and less similar sequences for PAM matrix; 30, 40, 70, 120, 250 and 62, 45, 50, 62, 80, 100 matrix versions are available for PAM and BLOSUM, respectively.

### Import the package first
   
```{r} 
# Import the package 
  library(kmeRs)
```

### Example 1. How to display PAM or BLOSUM matrix used for calculation?

Simply apply the **kmeRs_similarity_matrix** function and mark as an input the vector contains the nucleotides letters for witch the score should be calculated.    

```{r} 
# Simple BLOSUM62 similarity matrix for all DNA nucleotides
  result <- kmeRs_similarity_matrix(kmers_given = c("A", "T", "C", "G"), submat = "BLOSUM62")
# Fancy knitr table
  knitr::kable(result)
```

### Example 2. How to find the most *'different'* k-mer from the given set of k-mers?

In this example, the most *'different'* k-mer to "GATTACA" sequence will be indicated from given set of heptamers. Here, 7 heptamer (being an anagram of the movie title "GATTACA") are given, as follow:  

```{r} 
# Given hexamers
  kmers_given <- c("GATTACA", "ACAGATT", "GAATTAC", "GAAATCT", "CTATAGA", "GTACATA", "AACGATT")
# Matrix calculation 
  result <- kmeRs_similarity_matrix(kmers_given = c("GATTACA"), compare_to = kmers_given , submat = "BLOSUM62") 
# Fancy knitr table
  knitr::kable(result) 
```

Now, applying **kmeRs_score_and_sort** function the total score is calculated and the matrix is sorted by decreasing score value. The lowest value (in case of BLOSUM) indicates the most *'different'* sequence from given k-mers, in contrast to the highest value which indicates the most similar one. 

```{r} 
# Score and sort the matrix  
  result <- kmeRs_score_and_sort(result)
# Fancy knitr table
  knitr::kable(result)
```

As can be observed, the most *'different'* sequence to GATTACA is ACAGATT with total score equal to 1 and the most similar to GATTACA sequence is of course GATTACA sequence with the highest score equal to 37. 

### Example 3. How to find the most *'different'* k-mer to whole given set of k-mers? 

In this example, the most *'different'* k-mer to whole given set of heptamers will be indicated. The same heptamers as in example 2 are used. 

```{r} 
# Given hexamers
  kmers_given <- c("GATTACA", "ACAGATT", "GAATTAC", "GAAATCT", "CTATAGA", "GTACATA", "AACGATT")
# Matrix calculation 
  result <- kmeRs_similarity_matrix(kmers_given = kmers_given, submat = "BLOSUM62")
# Score the matrix and sort by decreasing score 
  result <- kmeRs_score_and_sort(result)
# Fancy knitr table
  knitr::kable(result)
  
```

As can be observed, the most *'different'* sequence to all given heptamers is CTATAGA with score equal to 62 and the most similar sequence is GAAATCT with the highest score equal to 97.

### Example 4. How to calculate basic statistics for the matrix?

Applying function **kmeRs_statistics** to the result matrix (here, result matrix from example 3) the basic statistics can be calculated as additional columns. When **summary_statistics_only** is set to TRUE only summary table is returned. It is much more elegant way to present results, especially in case of *'big data'* output.  

```{r}
# Score the matrix and sort by decreasing score 
  result <- kmeRs_statistics(result)
# Fancy knitr table
  knitr::kable(result[ , 1:(length(result[1, ])-4)])
```

## References
